bitcoin/src/pow.rs

```diff
@@ -0,0 +1,1530 @@
+        let hash = U256::from_le_bytes(hash.into_inner());
+        hash <= self.0
+    }
+
+    /// Converts this [`Target`] to [`Work`].
+    ///
+    /// "Work" is defined as the work done to mine a block with this target value (recorded in the
+    /// block header in compact form as nBits). This is not the same as the difficulty to mine a
+    /// block with this target (see `Self::difficulty`).
+    ///
+    /// # Examples
+    ///
+    /// ```
+    /// use bitcoin::Target;
+    /// let t = Target::from(0xdeadbeef_u32);
+    /// assert_eq!(t.to_work().to_target(), t)
+    /// ```
+    pub fn to_work(self) -> Work { Work(self.0.inverse()) }
+
+    /// Computes the popular "difficulty" measure for mining.
+    ///
+    /// Difficulty represents how difficult the current target makes it to find a block, relative to
+    /// how difficult it would be at the highest possible target (highest target == lowest difficulty).
+    ///
+    /// For example, a difficulty of 6,695,826 means that at a given hash rate, it will, on average,
+    /// take ~6.6 million times as long to find a valid block as it would at a difficulty of 1, or
+    /// alternatively, it will take, again on average, ~6.6 million times as many hashes to find a
+    /// valid block
+    ///
+    /// # Note
+    ///
+    /// Difficulty is calculated using the following algorithm `max / current` where [max] is
+    /// defined for the Bitcoin network and `current` is the current [target] for this block. As
+    /// such, a low target implies a high difficulty. Since [`Target`] is represented as a 256 bit
+    /// integer but `difficulty()` returns only 128 bits this means for targets below approximately
+    /// `0xffff_ffff_ffff_ffff_ffff_ffff` `difficulty()` will saturate at `u128::MAX`.
+    ///
+    /// [max]: Target::max
+    /// [target]: crate::blockdata::block::BlockHeader::target
+    pub fn difficulty(&self) -> u128 {
+        let d = Target::MAX.0 / self.0;
+        d.saturating_to_u128()
+    }
+}
+do_impl!(Target);
+
+impl Div for Target {
+    type Output = Self;
+    fn div(self, rhs: Self) -> Self { Self(self.0 / rhs.0) }
+}
+
+/// Encoding of 256-bit target as 32-bit float.
+///
+/// This is used to encode a target into the block header. Satoshi made this part of consensus code
+/// in the original version of Bitcoin, likely copying an idea from OpenSSL.
+#[derive(Copy, Clone, Debug, Default, PartialEq, Eq, PartialOrd, Ord, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct CompactTarget(u32);
+
+impl CompactTarget {
+    /// Creates a [`CompactTarget`] from a consensus encoded `u32`.
+    pub fn from_consensus(bits: u32) -> Self { Self(bits) }
+
+    /// Returns the consensus encoded `u32` representation of this [`CompactTarget`].
+    pub fn to_consensus(self) -> u32 { self.0 }
+}
+
+impl From<Target> for CompactTarget {
+    fn from(t: Target) -> Self { t.to_compact() }
```

Comment:
One potentially surprising property of this is that `x == Target::from(CompactTarget::from(x)` doesn't always hold.

Maybe we should call it `to_compact_lossy` and skip this `From` impl. Maybe also implement `TryFrom` returning error if any information would be lost (but that' seems too much).

Reply:
Good point, I've done as suggested and further commented `to_compact_lossy` also.